Move minimising functionality from bem make to the techs

scf2k opened the pull request.

Commit:
First version

Commit:
Pass minimize argument to borschik depending on YENV value

Commit:
exclude min.js from from getTechs()

arikon commented on lib/techs/v2/borschik.js:
```diff
@@ -0,0 +1,118 @@
+while(parent.parent) parent = parent.parent;
+
+// Check if this file is executed directly via fork() from create() method bellow.
+if (process.argv[1] === parent.filename) {
```
`if (require.main === module)`

arikon commented on lib/techs/v2/borschik.js:
```diff
@@ -0,0 +1,118 @@
+
+    getCreateResult: function(path, suffix, vars) {
+
+        var worker = cp.fork(__filename, [], {silent: true}),
```
Я бы всё добро про вызов воркера вынес в утилити функцию

arikon commented on lib/techs/v2/min.js.js:
```diff
@@ -0,0 +1,17 @@
+
+exports.techMixin = {
+
+    getSuffixes: function() {
```
Какие суффиксы здесь указываются, исходные? Почему название `getSuffixes()`, а не какое-нибудь `getSourceSuffixes()`?

scf2k replied:
Исходные. Как предусмотрено базовой технологией.

SevInf commented on lib/techs/v2/borschik.js:
```diff
@@ -0,0 +1,118 @@
+            };
+
+        worker.stdout.on('data', function(data) {
+            out += data.toString();
```
`toString` сам позовется.

SevInf commented on lib/techs/v2/borschik.js:
```diff
@@ -0,0 +1,118 @@
+     * 'js' will be returned otherwise.
+     */
+    getMinTech: function(sourceSuffix) {
+        if (('.' + sourceSuffix).match(/\.css$/)) return 'css';
```
`PATH.extname` же

Commit:
code clean up

SevInf commented on lib/techs/v2/min.js:
```diff
@@ -0,0 +1,102 @@
+     * 'js' will be returned otherwise.
+     */
+    getMinTech: function(sourceSuffix) {
+        return PATH.extname('dummy.' + sourceSuffix) === '.css'? 'css': 'js';
```
Зачем dummy?

diunko replied:
@SevInf Для примера?

scf2k replied:
потому что `extname('css')` вернет пустую строку

SevInf commented on lib/techs/v2/min.js:
```diff
@@ -0,0 +1,102 @@
+     */
+    getCreateSuffixes: function() {
+        return this.getSuffixes()
+            .map(function(suffix) {
```
Можно сократить до `.map(this.getMinSuffix, this)`

Commit:
Add missing return

arikon commented:
@scf2k Этот PR актуален?

scf2k commented:
@arikon да, готов к вливанию

Commit:
Optimize map call in min.js

arikon commented:
Переименуй его, пожалуйста

16.10.2013, 15:31, "Vladimir Alaev" notifications@github.com:

> @arikon да, готов к вливанию
> 
> —
> Reply to this email directly or view it on GitHub.

## 

Отправлено из мобильной Яндекс.Почты: http://m.ya.ru/ymail

arikon commented:
LGTM

@SevInf @diunko Будете смотреть?

diunko commented on lib/techs/v2/min.js:
```diff
@@ -0,0 +1,100 @@
+    getCreateResult: function(path, suffix, vars) {
+
+        return U.fork(
+            __filename,
```
Why use custom fork here? (And not use forked build)

scf2k replied:
it's a helper function

Commit:
Link to *.min.* files instead of _*.* in the bemjson files

Add techs in the make.js to build min.css, min.ie.css and min.18n.js,
fix tests

Commit:
Introduce preferFork() method.

Commit:
Remove BorschikNode class and borschik worker script

scf2k commented:
@SevInf @arikon @diunko посмотрите, переделал на встроенный форк

arikon commented:
Looks good to me

arikon commented on lib/tech/v1.js:
```diff
@@ -661,6 +661,15 @@ var Q = require('q'),
+         * @return {Boolean}
+         */
+        preferFork: function() {
         }
```
Мне совсем не нравятся функции с пустым телом, их хочется удалить (несмотря на то, что это stub). Я бы добавил комментарий в тело функции, явно описывающий, что это stub.

SevInf replied:
Или `return false`.

arikon commented on lib/tech/v2.js:
```diff
@@ -685,6 +685,15 @@ var Q = require('q'),
+         * @return {Boolean}
+         */
+        preferFork: function() {
         }
```
`// stub`

arikon commented on lib/util.js:
```diff
@@ -962,6 +963,28 @@ exports.exec = function(cmd, options, resolveWithOutput) {
 
 };
 
+exports.fork = function(cmd, args, msg) {
```
Не хватает jsdoc.

Эта функция где-то используется?

arikon commented on test/data/make/project/.bem/make.js:
```diff
@@ -15,8 +15,28 @@ module.exports = function(make) {
+                        'min.i18n.js': {
+                            baseTechPath: 'v2/min',
+                            getSuffixes: function() {
+                                return ['en.js', 'ru.js', 'js'];
```
Не лучше ли будет завязаться на переменную со списком языков?

arikon commented:
@scf2k Вова, поправь по последним замечаниям и сделай rebase со squash для коммитов, которые вносят изменения в одни и те же места.

И вливаем.

Commit:
Add comments/jsdoc; min.i18n.js in the tests uses BEM_I18N_LANGS env for suffixes array

scf2k commented:
@arikon @diunko @SevInf поправил согласно комментариям

SevInf commented:
lgtm

Commit:
Fix min.js.js after commits reordering

Commit:
Tests: add ie.css to min.ie.css dependencies.
